File: alat/clean-schema.rs
use std::env;
use std::fs;
use std::path::Path;
use regex::Regex;

fn main() {
    // 1. Ambil folder dari argument atau default ke './artikelx/'
    let args: Vec<String> = env::args().collect();
    let target_folder = args.get(1).map(|s| s.as_str()).unwrap_or("./artikelx/");

    // 2. Regex untuk ld+json (Case Insensitive & Multiline diatur via flag (?is))
    // (?is) -> i: case-insensitive, s: allow . to match newline (\n)
    let schema_regex = Regex::new(r"(?is)<script\s+type=[\x22\x27]application/ld\+json[\x22\x27]>.*?</script>").unwrap();

    let path = Path::new(target_folder);
    if !path.exists() {
        eprintln!("❌ Folder tidak ditemukan: {}", target_folder);
        return;
    }

    println!("🧹 Memulai pembersihan STRICT LOCAL (RUST) di: {:?}", path.canonicalize().unwrap_or(path.to_path_buf()));
    println!("⚠️  Non-Recursive Mode: Sub-folder tidak akan disentuh.");
    println!("{}", "-".repeat(50));

    let mut count_cleaned = 0;
    let mut count_total = 0;

    // 3. Baca isi folder (Strict Local)
    if let Ok(entries) = fs::read_dir(path) {
        for entry in entries.flatten() {
            let file_path = entry.path();
            
            // Pastikan hanya file dan berakhiran .html / .htm
            if file_path.is_file() {
                let ext = file_path.extension().and_then(|s| s.to_str()).unwrap_or("").to_lowercase();
                if ext == "html" || ext == "htm" {
                    count_total += 1;
                    
                    if let Ok(content) = fs::read_to_string(&file_path) {
                        if schema_regex.is_match(&content) {
                            let cleaned_content = schema_regex.replace_all(&content, "");
                            if fs::write(&file_path, cleaned_content.as_ref()).is_ok() {
                                println!("✅ Cleaned: {:?}", file_path.file_name().unwrap());
                                count_cleaned += 1;
                            }
                        }
                    }
                }
            }
        }
    }

    println!("{}", "-".repeat(50));
    println!("📊 HASIL PEMBERSIHAN (RUST VERSION)");
    println!("📂 Total file HTML ditemukan : {}", count_total);
    println!("✨ File yang dibersihkan      : {}", count_cleaned);
    println!("😴 File sudah bersih          : {}", count_total - count_cleaned);
    println!("{}", "-".repeat(50));
}

File: dapur/XXX/clean-schema.js
// cleaner.js
import { readdir, stat } from "node:fs/promises";
import { join, resolve, extname } from "node:path";

// ======================================================
// KONFIGURASI PEMBERSIH (OPTIMIZED FOR BUN)
// ======================================================
const TARGET_FOLDER = Bun.argv[2] || './artikelx/';
const SCHEMA_REGEX = /<script\s+type="application\/ld\+json">[\s\S]*?<\/script>/gi;
const EXTENSIONS = [".html", ".htm"];

async function cleanSchemaInFile(filePath) {
    try {
        const file = Bun.file(filePath);
        const content = await file.text();

        // Pakai Regex.test dulu biar nggak perlu write kalau nggak ada perubahan
        if (SCHEMA_REGEX.test(content)) {
            const cleanedContent = content.replace(SCHEMA_REGEX, "");
            
            // Bun.write jauh lebih cepat dibanding fs.writeFileSync
            await Bun.write(filePath, cleanedContent);
            return true;
        }
        return false;
    } catch (err) {
        console.error(`❌ Error pada ${filePath}: ${err.message}`);
        return false;
    }
}

async function runCleanerLocal() {
    const absolutePath = resolve(TARGET_FOLDER);
    
    // Cek folder pakai Bun.file().exists()
    if (!(await Bun.file(absolutePath).exists) && (await stat(absolutePath).catch(() => ({ isDirectory: () => false }))).isDirectory() === false) {
        // Fallback check simple
        try { await readdir(absolutePath) } catch {
            console.error(`❌ Folder tidak ditemukan atau bukan direktori: ${absolutePath}`);
            return;
        }
    }

    console.log(`\n🧹 Memulai pembersihan STRICT LOCAL (Bun Mode) di: ${absolutePath}`);
    console.log("⚠️  Non-Recursive Mode: Sub-folder tidak akan disentuh.");
    console.log("-".repeat(50));
    
    let countCleaned = 0;
    let countTotal = 0;

    // Ambil semua file di folder tersebut
    const items = await readdir(absolutePath);

    // Kita pakai Promise.all agar prosesnya paralel (Concurrent), bukan satu-satu (Sequential)
    // Ini yang bikin Bun makin berasa ngebutnya.
    const tasks = items.map(async (item) => {
        const itemPath = join(absolutePath, item);
        const itemStat = await stat(itemPath);

        const isHtml = EXTENSIONS.includes(extname(item).toLowerCase());

        if (itemStat.isFile() && isHtml) {
            countTotal++;
            const wasCleaned = await cleanSchemaInFile(itemPath);
            if (wasCleaned) {
                console.log(`✅ Cleaned: ${item}`);
                countCleaned++;
            }
        }
    });

    await Promise.all(tasks);

    console.log("-".repeat(50));
    console.log(`📊 HASIL PEMBERSIHAN (BUN OPTIMIZED)`);
    console.log(`📂 Total file HTML ditemukan : ${countTotal}`);
    console.log(`✨ File yang dibersihkan      : ${countCleaned}`);
    console.log(`😴 File sudah bersih          : ${countTotal - countCleaned}`);
    console.log("-".repeat(50));
}

runCleanerLocal();
